sensorium/training/trainers.py

```diff
@@ -168,6 +260,43 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
         lr_decay_steps=lr_decay_steps,
     ):
 
+        if include_kldivergence and epoch == dec_starting_epoch:
+            # TODO: include hidden dimension
+            kmeans = KMeans(
+                n_clusters=cluster_number, n_init=kmeans_init, random_state=seed
+            )
+            feature_list = []
+            features_subset = []
+            random_indices = {}
+            # form initial cluster centres
+            with torch.no_grad():
+                for k, readout in model.readout.items():
+                    features = readout.features.cpu().detach().squeeze().T.numpy()
+                    feature_list.append(np.array(features))
+                    # TODO remove subset
+                    rng = np.random.default_rng(seed)
+                    random_index = rng.choice(
+                        features.shape[0], size=int(subsamples), replace=False
+                    )
+                    random_indices[k] = random_index
+                    features_subset.append(features[random_indices[k], :])
+                features = np.vstack(feature_list)
+                features_subset = np.vstack(features_subset)
+                predicted = kmeans.fit_predict(features)
+
+                # predicted_previous = torch.tensor(np.copy(predicted), dtype=torch.long)
+                # _, accuracy = cluster_accuracy(predicted, actual.cpu().numpy())
+                cluster_centers = torch.tensor(
+                    kmeans.cluster_centers_, dtype=torch.float, requires_grad=True
+                )
+                cluster_centers = cluster_centers.to(device, non_blocking=True)
```

Comment:
I am not sure if these lines should be inside `with torch.no_grad():`, I actually think they should be outside like 
```
with torch.no_grad():
    for k, readout in model.readout.items():
        features = readout.features.cpu().detach().squeeze().T.numpy()
        ....
cluster_centers = torch.tensor(
    kmeans.cluster_centers_, dtype=torch.float, requires_grad=True
)
cluster_centers = cluster_centers.to(device, non_blocking=True)

```